pkg/frontend/querymiddleware/split_by_interval_test.go

```diff
@@ -0,0 +1,161 @@
+		}
+
+		series = append(series, newSeries(newTestCounterLabels(seriesID), start.Add(-lookbackDelta), end, step, gen))
+		seriesID++
+	}
+
+	// Add a special series whose data points end earlier than the end of the queried time range
+	// and has NO stale marker.
+	series = append(series, newSeries(newTestCounterLabels(seriesID),
+		start.Add(-lookbackDelta), end.Add(-5*time.Minute), step, factor(2)))
+	seriesID++
+
+	// Add a special series whose data points end earlier than the end of the queried time range
+	// and HAS a stale marker at the end.
+	series = append(series, newSeries(newTestCounterLabels(seriesID),
+		start.Add(-lookbackDelta), end.Add(-5*time.Minute), step, stale(end.Add(-6*time.Minute), end.Add(-4*time.Minute), factor(2))))
+	seriesID++
+
+	// Add a special series whose data points start later than the start of the queried time range.
+	series = append(series, newSeries(newTestCounterLabels(seriesID),
+		start.Add(5*time.Minute), end, step, factor(2)))
+	seriesID++
+
+	// Add histogram series.
+	for i := 0; i < numHistograms; i++ {
+		for bucketIdx, bucketLe := range histogramBuckets {
+			// We expect each bucket to have a value higher than the previous one.
+			gen := factor(float64(i) * float64(bucketIdx) * 0.1)
+			if i >= numHistograms-numStaleHistograms {
+				// Wrap the generator to inject the staleness marker between minute 10 and 20.
+				gen = stale(start.Add(10*time.Minute), start.Add(20*time.Minute), gen)
+			}
+
+			series = append(series, newSeries(newTestHistogramLabels(seriesID, bucketLe),
+				start.Add(-lookbackDelta), end, step, gen))
+		}
+
+		// Increase the series ID after all per-bucket series have been created.
+		seriesID++
+	}
+
+	// Create a queryable on the fixtures.
+	queryable := storageSeriesQueryable(series)
+
+	for testName, testData := range tests {
+		// Change scope to ensure it work fine when test cases are executed concurrently.
+		testData := testData
+
+		t.Run(testName, func(t *testing.T) {
+			t.Parallel()
+			reqs := []Request{
+				&PrometheusInstantQueryRequest{
+					Path:  "/query",
+					Time:  util.TimeToMillis(end),
+					Query: testData.query,
+				},
+			}
+
+			for _, req := range reqs {
+				t.Run(fmt.Sprintf("%T", req), func(t *testing.T) {
+					engine := newEngine()
+					downstream := &downstreamHandler{
+						engine:    engine,
+						queryable: queryable,
+					}
+
+					// Run the query with the normal engine
+					expectedRes, err := downstream.Do(context.Background(), req)
+					require.Nil(t, err)
+					expectedPrometheusRes := expectedRes.(*PrometheusResponse)
+					sort.Sort(byLabels(expectedPrometheusRes.Data.Result))
+
+					// Ensure the query produces some results.
+					require.NotEmpty(t, expectedPrometheusRes.Data.Result)
+					requireValidSamples(t, expectedPrometheusRes.Data.Result)
+
+					splittingware := newSplitByIntervalMiddleware(true, 1*time.Minute, mockLimits{}, log.NewNopLogger(), engine)
```

Comment:
1m split interval looks very very low. Some bugs may not surface with that. It could be more interesting running these test cases with different split intervals.

Reply:
As discussed via slack, this will be addressed in a separate PR and tracked in issue: https://github.com/grafana/mimir-squad/issues/774